pkg/kube/krt/processor.go

```diff
@@ -0,0 +1,251 @@
+	l.send(initialEvents, true)
+	return l.Synced()
+}
+
+func (o *handlerSet[O]) Distribute(events []Event[O], initialSync bool) {
+	o.mu.RLock()
+	defer o.mu.RUnlock()
+	for _, listener := range o.handlers {
+		listener.send(slices.Clone(events), initialSync)
+	}
+}
+
+// Synced returns a Syncer that will wait for all registered handlers (at the time of the call!) are synced.
+func (o *handlerSet[O]) Synced() Syncer {
+	o.mu.RLock()
+	syncer := multiSyncer{syncers: make([]Syncer, 0, len(o.handlers))}
+	for _, listener := range o.handlers {
+		syncer.syncers = append(syncer.syncers, listener.Synced())
+	}
+	o.mu.RUnlock()
+	return syncer
+}
+
+// processorListener is a fork of the upstream Kubernetes one. It has minimal tweaks to fit into
+// our use case better, but the most important one being that processorListener is private in Kubernetes.
+//
+// processorListener relays notifications from a sharedProcessor to
+// one ResourceEventHandler --- using two goroutines, two unbuffered
+// channels, and an unbounded ring buffer.  The `add(notification)`
+// function sends the given notification to `addCh`.  One goroutine
+// runs `pop()`, which pumps notifications from `addCh` to `nextCh`
+// using storage in the ring buffer while `nextCh` is not keeping up.
+// Another goroutine runs `run()`, which receives notifications from
+// `nextCh` and synchronously invokes the appropriate handler method.
+//
+// processorListener also keeps track of the adjusted requested resync
+// period of the listener.
+type processorListener[O any] struct {
+	nextCh chan any
+	addCh  chan eventSet[O]
+
+	handler func(o []Event[O], initialSync bool)
+
+	syncTracker *countingTracker
+
+	// pendingNotifications is an unbounded ring buffer that holds all notifications not yet distributed.
+	// There is one per listener, but a failing/stalled listener will have infinite pendingNotifications
+	// added until we OOM.
+	// TODO: This is no worse than before, since reflectors were backed by unbounded DeltaFIFOs, but
+	// we should try to do something better.
+	pendingNotifications buffer.RingGrowing
+}
+
+func newProcessListener[O any](
+	handler func(o []Event[O], initialSync bool),
+	upstreamSyncer Syncer,
+) *processorListener[O] {
+	bufferSize := 1024
+	ret := &processorListener[O]{
+		nextCh:               make(chan any),
+		addCh:                make(chan eventSet[O]),
+		handler:              handler,
+		syncTracker:          &countingTracker{upstreamHasSynced: upstreamSyncer, synced: make(chan struct{})},
+		pendingNotifications: *buffer.NewRingGrowing(bufferSize),
+	}
+
+	return ret
+}
+
+type eventSet[O any] struct {
+	event           []Event[O]
+	isInInitialList bool
+}
+
+func (p *processorListener[O]) send(event []Event[O], isInInitialList bool) {
+	if isInInitialList {
+		// The initial sync had no items to process, so we would never call syncTracker.Finish().
+		// Just directly mark it as done here
+		if len(event) == 0 {
+			close(p.syncTracker.synced)
```

Comment:
This may be straight from upstream and we just keep it as if for that reason but rather than this close which is functional but seems out of place. Should we just have a receiver function for syncTracker like `MarkFinished()` which can handle this?